singularity-eos/eos/eos_gruneisen.cpp

```diff
@@ -12,12 +12,155 @@
 // publicly and display publicly, and to permit others to do so.
 //------------------------------------------------------------------------------
 
+#include <sstream>
+
+#include <root-finding-1d/root_finding.hpp>
 #include <singularity-eos/eos/eos.hpp>
 
 namespace singularity {
 
 PORTABLE_INLINE_FUNCTION Real square(const Real x) { return x * x; }
 PORTABLE_INLINE_FUNCTION Real cube(const Real x) { return x * x * x; }
+PORTABLE_INLINE_FUNCTION Real find_min_bounded_val(const Real root1, const Real root2,
+                                                   const Real min, const Real max) {
+  // Try to find the minimum bounded value. If none exists, return a value less than the
+  // minimum
+  const Real minroot = std::min(root1, root2);
+  const Real maxroot = std::max(root1, root2);
+  if (minroot > min && minroot < max) {
+    return minroot;
+  } else if (maxroot > min && maxroot < max) {
+    return maxroot;
+  } else {
+    return min - 1.;
+  }
+}
+/*
+The Gruneisen EOS diverges at a specific compression. Ensure that the maximum density is
+below the smallest singularity in the reference pressure curve
+*/
+PORTABLE_FUNCTION Real Gruneisen::GetRhoMax(const Real s1, const Real s2, const Real s3,
+                                            const Real rho0) {
+  // Polynomial from the denominator of the reference pressure curve:
+  auto poly = [=](Real eta) { return 1 - s1 * eta - s2 * square(eta) - s3 * cube(eta); };
+
+  // First find the eta root. A negative root indicates that there is no maximum density.
+  Real root = 0; // Non-sensical root means a root hasn't been found yet
+  Real discriminant, root1, root2;
+  if (s2 == 0 && s3 == 0 && s1 > 0) {
+    // Linear Us-up analytic root
+    root = 1 / s1;
+  } else if (s3 == 0 && s2 != 0) {
+    // Quadratic Us-up
+    discriminant = square(s1) + 4 * s2;
+    if (discriminant < 0) {
+      root = -1; // Imaginary roots, so no limit
+    } else {
+      root1 = (s1 + std::sqrt(square(s1) + 4 * s2)) / (-2 * s2);
+      root2 = (s1 - std::sqrt(square(s1) + 4 * s2)) / (-2 * s2);
+      root = find_min_bounded_val(root1, root2, 0, 1);
+    }
+  } else if (s3 != 0) {
+    // Cubic Us-up
+    discriminant = -18 * s3 * s2 * s1 - 4 * cube(-s2) + square(s2) * square(s1) -
+                   4 * s3 * cube(s1) - 27 * square(s3);
+    // Use discriminant to help find roots
+    if (discriminant == 0) {
+      // Easy analytical roots (probably not the case)
+      if (square(s2) == 3 * s3 * s1) {
+        // Single root with multiplicity of 3 (inflection point is root)
+        // Also guards against divide by zero
+        root = -s2 / 3 / s3;
+      } else {
+        // One root has multiplicity of 2 (local extremum is a root)
+        root1 = (-9 * s3 - s2 * s1) / 2 / (square(s2) - 3 * s3 * s1);
+        root2 = (-4 * s3 * s2 * s1 - 9 * square(s3) + cube(s2)) /
+                (-s3 * (square(s2) - 3 * s3 * s1));
+        root = find_min_bounded_val(root1, root2, 0, 1);
+      }
```

Comment:
Given that this only gets called in initialization, is there a reason to do the analytic root finding at all? Maybe we could save some code and always use the numerical root finder?

If we did the numerical root finding always, we could use the analytic formulae as checks that everything is working properly in the unit tests.

Reply:
The reason I'm doing analytic root finding is because I need to guarantee that the  minimum root is found. I briefly looked into the numerics of this and decided that there was no guaranteed way to find the _minimum_ bounded root without more information.

Feel free to find a better way though.

Reply:
Are you talking about the cubic case specifically? I think the extrema equations are still valid when the discriminant is also zero so I could probably relax the condition to `if (discriminant >= 0) {` and eliminate that code. The good thing about writing that code is that I do at least test it so the tests should automatically tell me whether it's feasible to eliminate the code.

For the quadratic and linear cases I don't think you're saving much. For the quadratic case you'd still need to find the extremum anyway and then have a conditional for the sign of `s2` to determine whether you're greater than or less than the extremum.

Reply:
Ah yeah I guess the extremum trick only works for cubic.

Reply:
I added some comments to the extremum trick to make it more apparent that the trick works since the roots must be in a region of the polynomial that is _decreasing_. This is because the polynomial is pinned to 1 at $\eta = 0$, so there must be a decreasing region between $\eta = 0$ and $\eta = 1$ for there to be a possibility of any roots.

I also realized another reason to delete the analytic root part: the roots will only have multiplicity (i.e. the determinant is zero) when an extremum _**is**_ a root anyway. So I'm now requiring an explicit sign change across the bounds and otherwise testing the bounds themselves since the iterative approaches have a hard time converging if the root is an extremum.

Reply:
That sounds good! We should double check it all works in the tests, of course. We can use the code you just replaced as a sanity check. :)

Reply:
I believe I already have tests that capture the behavior of the previous code so in that sense I think it works.

Interestingly the tests for root multiplicity (i.e. roots are extrema) failed when using the iterative root find because the tolerances were too tight. The main issue is that the root find tolerance is too loose for this particular situation where the dy is basically zero. I would have assumed that regula falsi would converge directly to a bound if that bound is a root though, so I'm not quire sure what's going on.

Either way, skipping the root find when one of the bounds is a solution is probably the intelligent thing to do here anyway.

Reply:
Yeah that seems fine. My experience with Regula Falsi has been it's a little sensitive compared to bisection or secant actually. It's very fast, but it seems to fail out more frequently.